packages/web/app/src/lib/preflight-sandbox/graphiql-plugin.tsx

```diff
@@ -257,16 +286,20 @@ export function usePreflightScript(args: {
         }
 
         if (ev.data.type === IFrameEvents.Outgoing.Event.result) {
-          const mergedEnvironmentVariables = JSON.stringify(
-            {
-              ...safeParseJSON(latestEnvironmentVariablesRef.current),
-              ...ev.data.environmentVariables,
-            },
-            null,
-            2,
-          );
-          setEnvironmentVariables(mergedEnvironmentVariables);
-          latestEnvironmentVariablesRef.current = mergedEnvironmentVariables;
+          const mergedEnvironmentVariablesEncoded = Kit.JSON.encodePretty({
+            ...decodeEnvironmentVariables(latestEnvironmentVariablesRef.current),
+            ...ev.data.environmentVariables,
+          });
+          setEnvironmentVariables(mergedEnvironmentVariablesEncoded);
+          latestEnvironmentVariablesRef.current = mergedEnvironmentVariablesEncoded;
+
+          const mergedHeadersEncoded = Kit.JSON.encodePretty([
+            ...decodeHeaders(latestHeadersRef.current),
```

Comment:
This code copies the pattern that was in use for environment variables. I do not understand it. Specifically, when and why would the executed script append to an existing value in local storage rather than be expected to produce the total value upon every execution?

Reply:
I think the using spread on the object like the environmentVariables is fine because if it's called multiple times then the keys will override the prior value. This change to headers may be an issue though because you are spreading an array. There may be a bug where `ev.data.headers` keeps getting appended to the list of headers unless that's deduped elsewhere.

Reply:
> This change to headers may be an issue though because you are spreading an array. There may be a bug where ev.data.headers keeps getting appended to the list of headers unless that's deduped elsewhere.

Great call out. Oddly enough it is deduped later but in a way that I consider a bug wherein multiple header values could be lost.

On the other hand, I agree with you of course that here we could have kind of the opposite problem where we turn what were single-value headers into duplicated/multi-value headers.

In order to solve this, I would like to understand the system better. Why/when are there local storage values merged here.

Reply:
I've come to better understand the system. Local storage is not needed in the first place for headers.